Fix: Blackpill-F4 and BMP DFU

ALTracer opened the pull request:
## Detailed description

* Stock STM32F4 BootROM USB DFU implementation is slow and doesn't leave DFU mode on warm resets or detach requests.
* BMP codebase has a dfu_f4 implementation, let's make it available as a faster alternative to BootROM DFU. (~2.4x faster uploads on local testing, performance increased from 8KiB/s to 18KiB/s). Changes to makefile, .text offset of 16k, of which 8k used. This bootloader is not mandatory.
* `blackpill-f4` inherited the usbdfu.c from `f4-discovery` with wrong clocks (different HSE 25/8, different core clock 84/96/168), let's stick with 96MHz
* I also took the liberty to pull-up the floating PA0 button for robustness. Now the KEY press on boot actually matters. However, LED_BOOTLOADER which is mapped to PC15 is not the sole user LED on PC13.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`) -- doesn't break
* [x] It builds as BMDA (`make PROBE_HOST=hosted`) -- not applicable
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues

Not closes per se, but I'd like to mention #1152 from which I took the idea. That was abandoned less than a year ago, so before the big blackpill-f4 migration by @lenvm. Note that the DFU flash sectors' description string is composed at runtime, and the first 16K sector appears read-only (not actually write-protected, though).

## Problems

Seeing as there are F401 users, we need to decide on a method of selecting RCC configurations from libopencm3, maybe a macro. F401 will run at 84Mhz with 2ws, F411 at 96Mhz with 3ws (90nm 30Mhz Flash with ART). Not 100 because of USB48. The core is decently powerful, I wouldn't want to limit it especially since the three platform variants have separate headers and linkerscripts.

Upon unlocking the `BMP_BOOTLOADER=1` DFU code I faced weird issues with USBOTG DWC core, like the issues I had when building NuttX for this board. On one patch variant it fails to set up address from PC, on another it doesn't appear at all. Before the patch both BMP firmware and ST BootROM worked and enumerated immediately, after the patch maybe only the DFU works and nothing else can use the DWC USB on jump_to_application or soft-reset into ST BootROM. I have to physically unplug the board and remove power from external debugger. I think it has to do with VBus B-session sensing, PA9/PA10 AF mux and proper reset between consecutive `usbd_init()`'s (or upon leaving dfu_f4) instead of simply adjusting SP and jumping to application.

I need advice from the reviewers on which `#if 0`-hidden code blocks to drop for good, I simply kept these for context.

dragonmux commented on src/platforms/common/blackpill-f4/blackpill-f4.c:
```diff
@@ -67,15 +74,23 @@ void platform_init(void)
 	rcc_periph_clock_enable(RCC_OTGFS);
 	rcc_periph_clock_enable(RCC_CRC);
 
+#if 0
```
Please do not commit commented out code. If this is not required, just delete it - if we need it back, that's what Git history is for.

dragonmux commented on src/platforms/common/blackpill-f4/usbdfu.c:
```diff
@@ -68,6 +72,13 @@ int main(void)
 	dfu_protect(false);
 	dfu_init(&USB_DRIVER);
+
+#if 0
```
Please do not commit commented out code

dragonmux reviewed (changes requested):
Principally this looks good. There are a few items that need fixing before we can consider merging this, but they're easy enough to fix.

Thank you for the contribution!

ALTracer commented on src/platforms/blackpill-f401cc/platform.h:
```diff
@@ -25,6 +25,8 @@
 
 #define PLATFORM_IDENT "(BlackPill-F401CC) "
 
+#define RCC_CLOCK_SCALE RCC_CLOCK_3V3_84MHZ
```
Renaming and removing newlines makes clang-format touch the adjacent line (indented with spaces not tabs). Done.

dragonmux replied:
The whitespace changes it causes are absolutely not a problem - GitHub makes the display of that easy to understand during review. We will always prefer more readable code over minimising the number of formatting changes.

dragonmux commented:
We need to check through one or two things in the makefile changes for sanity, but this looks much better. Please apply your cleanup commit as a fixup to the commit that introduced those `#if 0` blocks and rebase. If you're happy that the code presented here works even if it's not perfect, we'd be interested in getting this merged.

ALTracer commented:
I rebased the entire branch, re-shuffling commits with respective cleanups/fixups, slightly editing commit contents.

I decided to drop the ENABLE_DEBUG & rdimon change from this patchset. That stuff needs revisiting in coordination with #1507. Testing the change on my blackpill didn't solve the probe hanging on initalization, moreover the upstream bluepill debugging up hung too (don't know where, it remaps its SWJ-DP).

Also, I incorporated the change placing bootloader magic into the `.noinit` section implemented by libopencm3 linkerscripts. `arm-none-eabi-nm` shows the same symbol of `20000000 B magic` in both src/blackmagic.elf and src/blackmagic_dfu. This allowed me to drop a bunch of `#pragma`s which sole existence was caused by `*magic=&_ebss` trickery.

The original branch and a few unrelated changes on top of it will stay in my fork for a while.

dragonmux reviewed (commented):
With the bootloader for these platforms fixed, this is looking much better. Thank you! We have just one review comment on this new code and with that fixed it'll be ready for merge.

ALTracer commented on src/platforms/common/blackpill-f4/usbdfu.c:
```diff
@@ -22,40 +22,38 @@
 		magic[0] = 0;
 		magic[1] = 0;
-	} else
+	} else {
```
I added them because (I touch adjacent code and) they are necessary per #428 and https://www.kernel.org/doc/html/latest/process/coding-style.html#placing-braces-and-spaces , or are you not following this style guide anymore? There's no document on style guide in BMP repo, but `src/target/stm32l4.c:stm32l4_attach()` exhibits both styles.

dragonmux replied:
No, and we haven't been for ages. #428 is an old PR (pre 1b² taking the reins on the project) and our advice is consistent with the last 2 years of development.

Files exhibiting both styles is because it's been incrementally applied and if we or someone else has had no reason to touch lines, they haven't yet been updated. Additionally, if you look at the history of that specific file, you'll see we removed those braces ourself when we did the clang-tidy cleanup on that section of that file.

dragonmux replied:
We will note that the website incorrectly points to the kernel style, however the needs of this project diverged from the Linux kernel's, so that needs updating. Do note that the website (which the repo does link to) supplements in-repository documentation just like the Linux kernel website does for the kernel.

ALTracer replied:
Braces change reverted.
Linux kernel website documentation is generated from in-tree sources, and I couldn't find the BM website sources in this repo. Maybe they're in a neighbouring repo in this GH org?...

dragonmux replied:
[yes](https://github.com/blackmagic-debug/black-magic-org), the one clearly labelled as such

dragonmux reviewed (approved):
LGTM, merging as soon as the build completes
